base/src/main/java/uk/ac/ebi/atlas/search/analyticsindex/FullAnalyticsSearchService.java

```diff
@@ -0,0 +1,251 @@
+    private ImmutableMap<String, Double> parseMaximumExpressionOverNonSelectedAssayGroups(String json) {
+        ImmutableMap.Builder<String, Double> maxExpressionByGene = ImmutableMap.builder();
+
+        Collection<Map<String, Object>> buckets = JsonPath.read(json, "$.facets.genes.buckets.*");
+
+        buckets.forEach(
+                bucket -> maxExpressionByGene.put((String) bucket.get("val"), (double) bucket.get("max_expression")));
+
+        return maxExpressionByGene.build();
+    }
+
+
+//    private TreeMultimap<Integer, Pair<String, Double>> parseAverageExpressionOverSelectedAssayGroups(String json) {
+//        TreeMultimap<Integer, Pair<String, Double>> genesBySpecificity = TreeMultimap.create();
+//
+//        Collection<Map<String, String>> buckets = JsonPath.read(json, "$.facets.genes.buckets.*");
+//
+//        buckets.forEach(
+//                bucket ->
+//                        genesBySpecificity.put(
+//                                Integer.parseInt(bucket.get("count")),
+//                                Pair.of(bucket.get("val"), Double.parseDouble(bucket.get("avg_expression")))));
+//
+//        return genesBySpecificity;
+//    }
+
+    // TODO consider adding a private class that represents this bucket for safe type deserialization (no casts needed)
+    private TreeMultimap<Integer, Pair<String, Double>> parseAverageExpressionOverSelectedAssayGroups(String json,
+                                                                                                      int limit) {
+        TreeMultimap<Integer, Pair<String, Double>> genesBySpecificity = TreeMultimap.create();
+
+        Collection<Map<String, Object>> buckets = JsonPath.read(json, "$.facets.genes.buckets.*");
+        Iterator<Map<String, Object>> bucketsIterator = buckets.iterator();
+
+        int specificity = 0;
+        while (bucketsIterator.hasNext()) {
+            Map<String, Object> nextBucket = bucketsIterator.next();
+
+            // We must read the full set of genes with same specificity (the most highly expressed may be anywhere)
+            if ((int) nextBucket.get("count") > specificity) {
+                specificity = (int) nextBucket.get("count");
+                if (genesBySpecificity.entries().size() >= limit) {
+                    // We have read enough buckets, get outta here
+                    break;
+                }
+            }
+
+            // Keep reading all the buckets with this specificity
+            genesBySpecificity.put(
+                    specificity,
+                    Pair.of((String) nextBucket.get("val"), ((Number) nextBucket.get("avg_expression")).doubleValue()));
+            // avg_expression comes some times as Double and some times as BigDecimal
+
+        }
+
+        return genesBySpecificity;
+    }
+
+    protected String fetchResponseAsString(String url, SolrQuery query) {
```

Comment:
This looks like you took it from somewhere, add comments where you took it form before you forget?

Reply:
Again, this is from `AnalyticsQueryClient`, which will absorb most of the general functionality from this class.